File: 00-arrays-hashing/03-49-GroupAnagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function (strs) {
  let map = new Map();
  let a = "a".charCodeAt();

  strs.map((word) => {
    counter = new Array(26).fill(0);
    for (let c of word) {
      counter[c.charCodeAt() - a]++;
    }
    let strCounter = counter.toString();
    if (map.has(strCounter)) {
      map.get(strCounter).push(word);
    } else map.set(strCounter, [word]);
  });
  return [...map.values()];
};

File: 00-arrays-hashing/03-49-GroupAnagrams.py
import collections
from typing import List


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        myMap = collections.defaultdict(list)
        for word in strs:
            counter = [0] * 26
            a = ord("a")
            for c in word:
                counter[ord(c) - a] += 1
            myMap[tuple(counter)].append(word) 
        return myMap.values()

print(Solution().groupAnagrams(["eat","tea","tan","ate","nat","bat"]))